### PR TITLE
Types removal -  deprecate types in _bulk apis

### PR DESCRIPTION
**Pull request by markharwood**
* Deprecate transport client APIs with document types.
* Deprecate REST endpoints that have types in URLs or bodies.

Some remaining work (e.g. `CRUDDocumentationIT.testBulk` code examples) depends on @mayya-sharipova 's changes to `IndexRequest` for the singular index apis.

### CODE REVIEW

**Comment by markharwood** [comment]
run gradle build tests

**Comment by markharwood** [comment]
run gradle build tests

**Comment by markharwood** [comment]
run gradle build tests

**Comment by markharwood** [comment]
run gradle build tests

**Comment by markharwood** [comment]
The last remaining test failure is a rolling upgrade test (IndexingIT) which looks to be failing when a 6.x node is given a types removal warning by a 7.x node.
I've opened https://github.com/elastic/elasticsearch/pull/37105 in the 6.x branch to address this which when comitted should hopefully give us a green build here.

**Review by jtibshirani — COMMENTED** [review]
What a huge effort!! I left a few preliminary comments as I got up to speed with the approach.

File: server/src/main/java/org/elasticsearch/rest/action/document/RestBulkAction.java
Comment (markharwood) [reply server/src/main/java/org/elasticsearch/rest/action/document/RestBulkAction.java]: If that's the convention, sure,  I can make that change

**Comment by markharwood** [comment]
run gradle build tests

File: server/src/main/java/org/elasticsearch/rest/action/document/RestBulkAction.java
Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/rest/action/document/RestBulkAction.java]: 👍

File: client/rest-high-level/src/test/java/org/elasticsearch/client/BulkRequestWithGlobalParametersIT.java
Comment (jtibshirani) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/BulkRequestWithGlobalParametersIT.java]: One more instance of `expectVersionSpecificWarnings` that can be updated to `expectWarnings`.

Comment (jtibshirani) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/BulkRequestWithGlobalParametersIT.java]: One more instance of `expectVersionSpecificWarnings` that can be updated to `expectWarnings`.

File: qa/full-cluster-restart/src/test/java/org/elasticsearch/upgrades/FullClusterRestartIT.java
Comment (jtibshirani) [inline qa/full-cluster-restart/src/test/java/org/elasticsearch/upgrades/FullClusterRestartIT.java]: Same thought here.

File: server/src/main/java/org/elasticsearch/action/index/IndexRequest.java
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: It might be useful to add a comment to these files `IndexRequest`, `UpdateRequest`, and `DeleteRequest`) explaining the different states of the type parameter, as it is not very obvious to a newcomer. I really should've done this in my initial PRs that began to modify these files, apologies for punting on that.

**Review by jtibshirani — COMMENTED** [review]
This is looking good to me! I will wait on the build results before hitting 'approve', in case another issue comes up or you'd like another pair of eyes on it (some of the changes around defaults in `IndexRequest` are a bit subtle).

**Comment by jtibshirani** [comment]
@elasticmachine run gradle build tests 1

File: server/src/test/java/org/elasticsearch/rest/action/document/RestBulkActionTests.java
Comment (jtibshirani) [inline server/src/test/java/org/elasticsearch/rest/action/document/RestBulkActionTests.java]: In other PRs we have added a dedicated test here to ensure we're emitting deprecation warnings. However, there are several integration tests in this PR that maked typed calls and check for deprecation warnings, so I don't think it's strictly necessary.

**Review by jtibshirani — APPROVED** [review]
Thanks @markharwood, this looks good to me. My only other thought is that it'd be good to update the PR description with an explanation of how bulk, index, update, and delete requests have changed (perhaps mentioning `defaultTypeIfNull`).

**Comment by markharwood** [comment]
run gradle build tests

File: server/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java]: It looks like CRUDDocumentationIT is using this deprecatd ctor still (amongst other). Since its code that is contained in the docs, should we move that to using the typeless constructor?

File: server/src/main/java/org/elasticsearch/action/DocWriteRequest.java
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/action/DocWriteRequest.java]: Can you briefly explain why this is added? Only to improve my understanding of the PR.

File: server/src/main/java/org/elasticsearch/action/bulk/BulkRequestBuilder.java
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/action/bulk/BulkRequestBuilder.java]: Could the other methods using "type" be deprecated as well?

File: server/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java]: Could the other methods using "type" be deprecated as well?

File: server/src/main/java/org/elasticsearch/action/index/IndexRequest.java
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: Any reason this is removed here but still validated in DeleteRequest?

File: server/src/main/java/org/elasticsearch/action/index/IndexRequest.java
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: Maybe add the same comment to the writeTo() above in DeleteRequest too. It helps.

File: server/src/main/java/org/elasticsearch/action/update/UpdateRequest.java
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/action/update/UpdateRequest.java]: Same here, would be nice to transfer the comment from the other writeTo() method.

**Review by cbuescher — COMMENTED** [review]
@markharwood I did a first quick pass ignoring the changes in test classes, will take a look at those also if needed

File: server/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java
Comment (markharwood) [reply server/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java]: I'm not sure - it's documentation showing how `index` and `type` can be associated with the BulkRequest and "inherited" by an IndexRequest objects that are missing these values.

I suppose it's a general point if we want documentation to include supported-yet-deprecated features? I can pull the docs/code if we think that's the right thing. Thoughts? cc @jtibshirani

File: server/src/main/java/org/elasticsearch/action/DocWriteRequest.java
Comment (markharwood) [reply server/src/main/java/org/elasticsearch/action/DocWriteRequest.java]: Bulk requests can specify a default index and type which is automatically filled-in on any contained Index/Update/DeleteRequests that are missing these properties. A form of inheritance.

This is the method where any blank choice of type is overridden with the BulkRequest's choice of type. This code became necessary after a different PR switched the Index/Update/DeleteRequests to default to `_doc` instead of null - all the bulk type inheritance tests failed because we couldn't differentiate between objects where the user had explicitly defined a choice of `_doc` and those where the user had not made an explicit choice. Here we have to revert back to holding a null for this setting by default.

File: server/src/main/java/org/elasticsearch/action/index/IndexRequest.java
Comment (markharwood) [reply server/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: No, I can't think why they should differ

File: server/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java
Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java]: We've been removing documentation for deprecated features, with the understanding that users can refer to the previous version's documentation (where the feature is not deprecated).

**Comment by cbuescher** [comment]
@markharwood rest of the PR (test changes) look good to me btw.

File: server/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java]: Thanks for the clarification, I think you are refering to https://www.elastic.co/guide/en/elasticsearch/client/java-rest/master/java-rest-high-document-bulk.html#CO102-1. I would be in favour of removing the mentioning of the type there as well and changing to the other ctor then.

File: server/src/main/java/org/elasticsearch/action/DocWriteRequest.java
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/action/DocWriteRequest.java]: Thanks for the explanation, think I got it.

**Review by cbuescher — APPROVED** [review]
@markharwood thanks for the update, LGTM.

**Pull request merged**
